### PR TITLE
Registry auth

### PR DESCRIPTION
**Pull request by ghost**
Creating a pull request just so it's easier to compare / comment

### CODE REVIEW

File: test/integration/docker_build_test.exs [hunk 1]
```diff
@@ -96,4 +96,11 @@ defmodule ExDockerBuild.Integration.DockerBuildTest do
       assert log =~ "STEP 6/6 : CMD [\"cat\", \"/myvol/greeting\"]"
     end
   end
+
+  describe "authentication" do
+    test "try authenticate an illegal user" do
```
Comment (ghost) [inline test/integration/docker_build_test.exs:101]: try to authenticate

**Comment by ghost** [comment]
>> Current status : 
>> 
>> The problem is we aren't able to get the identity token (please clarify whether this is when calling the /auth endpoint on docker hub or a private docker registry instance) because the registry isn't returning the token. 
>> 
>> What expected the registry to return is 
>> 
>> ```{"IdentityToken":"<some base64 encoded token>","Status":"Login Succeeded"}```
>> 
>> Instead we receive : 
>> 
>> ```{"IdentityToken":"","Status":"Login Succeeded"}```
>> 
>> An example of the curl command was given in https://github.com/docker/for-mac/issues/1431
>> 
>> ```
>> curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" -d '{"username": "", "password": "", "serveraddress": "https://index.docker.io/v1/"}' -X POST http:/1.26/auth
>> ```
>> 
>> https://github.com/moby/moby/issues/26000
>> 
>> Someone pointed out that they solve it passing the whole credentials as a 'X-REGISTRY-AUTH' header, for EVERY API call, the problem is this will change the whole structure of how the docker build works now.
>> 

Yeah this is a pain - I've investigated docker for mac and it's storing the raw password in the keychain - not using tokens anywhere - the docker API docs are wrong - perhaps create an issue on their side to update them ? 

Yes, suggested workaround from https://github.com/docker/for-mac/issues/1431#issue-214088420 is to set 'X-REGISTRY-AUTH' HTTP header with the base64 encoded JSON. 

```
{
"username": "string",
"password": "string",
"email": "string",
"serveraddress": "string"
}
```

I'd suggest to get started just push and pull private images from docker hub (I think they provide one private image by default). The database schema (and perhaps more besides) will need to be structured so that the option to allow private docker repositories isn't too much pain to implement in the future. 

FYI - using the docker command line the trick for pushing to a private docker repo is : 

```
docker tag rpmtools-centos6 127.0.0.1:5000/foo/rpmtools-centos6
docker push 127.0.0.1:5000/rpmtools-centos7:latest
```

Thoughts @filipevarjao @sescobb27 ?

**Comment by filipevarjao** [comment]
I'm using Linux, so is not only an issue related to mac.

**Comment by sescobb27** [comment]
yes, i was thinking in maybe implement a push function that takes a credentials map something like `%{username: "xxx", password: "xxx"}` or `%{email: "xxx", password: xxx}` along with an optional `serveraddress`.

For private repositories i will suggest for each task add `docker_username:string` `docker_password:encrypted` `docker_server_address:string` `docker_image_name` that way we can inject it to the push command and give it a new name. name can be something we can use as a template in the future for custom naming? thoughts?

**Comment by sescobb27** [comment]
I think filipe can help me to implement this, adding those attributes to the tool so he can get more familiar with ecto/migrations/views/etc and we can pair this one. thoughts? @bryanhuntesl @filipevarjao

**Comment by ghost** [comment]
> yes, i was thinking in maybe implement a push function that takes a credentials map something like `%{username: "xxx", password: "xxx"}` or `%{email: "xxx", password: xxx}` along with an optional `serveraddress`.

Sounds good, I'd make the server address obligatory though - you don't want to accidentally push user credentials to the wrong server - I believe the endpoint address for docker hub is "https://index.docker.io/v1/"  (correct me if wrong).

**Comment by sescobb27** [comment]
ok, so what about making the server address required but having an option to set index.docker.io/v1 as default (i wouldn't like to type the same address over and over again if using the default)

**Comment by ghost** [comment]
Default works fine. Agreed.

**Pull request closed without merging**